extension/opampextension/registry_test.go

```diff
@@ -0,0 +1,283 @@
+
+		sender, err := registry.Register(capabilityString)
+		require.Nil(t, sender)
+		require.ErrorIs(t, err, capabilityErr)
+		require.Len(t, registry.capabilityToMsgChannels, 0, "Setting capability failed, but callback ended up in the map anyways")
+	})
+}
+
+func TestRegistry_ProcessMessage(t *testing.T) {
+	t.Run("Calls registered callback", func(t *testing.T) {
+		capabilityString := "io.opentelemetry.teapot"
+		messageType := "steep"
+		mesageBytes := []byte("blackTea")
+		customMessage := &protobufs.CustomMessage{
+			Capability: capabilityString,
+			Type:       messageType,
+			Data:       mesageBytes,
+		}
+
+		client := mockCustomCapabilityClient{}
+
+		registry := newCustomCapabilityRegistry(zap.NewNop(), client)
+
+		sender, err := registry.Register(capabilityString)
+		require.NotNil(t, sender)
+		require.NoError(t, err)
+
+		registry.ProcessMessage(customMessage)
+
+		require.Equal(t, customMessage, <-sender.Message())
+	})
+
+	t.Run("Skips blocked message channels", func(t *testing.T) {
+		capabilityString := "io.opentelemetry.teapot"
+		messageType := "steep"
+		mesageBytes := []byte("blackTea")
+		customMessage := &protobufs.CustomMessage{
+			Capability: capabilityString,
+			Type:       messageType,
+			Data:       mesageBytes,
+		}
+
+		client := mockCustomCapabilityClient{}
+
+		registry := newCustomCapabilityRegistry(zap.NewNop(), client)
+
+		sender, err := registry.Register(capabilityString, WithMaxQueuedMessages(0))
+		require.NotNil(t, sender)
+		require.NoError(t, err)
+
+		// If we did not skip sending on blocked channels, we'd expect this to never return.
+		registry.ProcessMessage(customMessage)
+
+		require.Equal(t, 0, len(sender.Message()))
+	})
+
+	t.Run("Callback is called only for its own capability", func(t *testing.T) {
+		teapotCapabilityString1 := "io.opentelemetry.teapot"
+		coffeeMakerCapabilityString2 := "io.opentelemetry.coffeeMaker"
+
+		messageType1 := "steep"
+		messageBytes1 := []byte("blackTea")
+
+		messageType2 := "brew"
+		messageBytes2 := []byte("blackCoffee")
+
+		customMessageSteep := &protobufs.CustomMessage{
+			Capability: teapotCapabilityString1,
+			Type:       messageType1,
+			Data:       messageBytes1,
+		}
+
+		customMessageBrew := &protobufs.CustomMessage{
+			Capability: coffeeMakerCapabilityString2,
+			Type:       messageType2,
+			Data:       messageBytes2,
+		}
+
+		client := mockCustomCapabilityClient{}
+
+		registry := newCustomCapabilityRegistry(zap.NewNop(), client)
+
+		teapotSender, err := registry.Register(teapotCapabilityString1)
+		require.NotNil(t, teapotSender)
+		require.NoError(t, err)
+
+		coffeeMakerSender, err := registry.Register(coffeeMakerCapabilityString2)
+		require.NotNil(t, coffeeMakerSender)
+		require.NoError(t, err)
+
+		registry.ProcessMessage(customMessageSteep)
+		registry.ProcessMessage(customMessageBrew)
+
+		require.Equal(t, customMessageSteep, <-teapotSender.Message())
+		require.Equal(t, customMessageBrew, <-coffeeMakerSender.Message())
```

Comment:
Can we check the channels are empty after we take out each message? Just to ensure it's not an ordering error and there's actually another message sitting in there.

Reply:
Yeah, good call, I've added checks that the channels are empty.